mlir/test/Quantum/DisentangleCNOTTest.mlir

```diff
@@ -0,0 +1,217 @@
+
+
+    %17:2 = quantum.custom "CNOT"() %MINUS_0, %MINUS_1 : !quantum.bit, !quantum.bit
+    %user_17:2 = quantum.custom "CRZ"(%cst) %17#0, %17#1 : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.custom "CNOT"
+    // CHECK: [[mid_17:%.+]] = quantum.custom "PauliZ"() [[MINUS_0]] : !quantum.bit
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[mid_17]], [[MINUS_1]] : !quantum.bit, !quantum.bit
+
+
+    %18:2 = quantum.custom "CNOT"() %MINUS_0, %ZERO_1 : !quantum.bit, !quantum.bit
+    %user_18:2 = quantum.custom "CRZ"(%cst) %18#0, %18#1 : !quantum.bit, !quantum.bit
+    // CHECK: [[_18:%.+]]:2 = quantum.custom "CNOT"() [[MINUS_0]], [[ZERO_1]] : !quantum.bit, !quantum.bit
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[_18]]#0, [[_18]]#1 : !quantum.bit, !quantum.bit
+
+
+    %19:2 = quantum.custom "CNOT"() %MINUS_0, %ONE_1 : !quantum.bit, !quantum.bit
+    %user_19:2 = quantum.custom "CRZ"(%cst) %19#0, %19#1 : !quantum.bit, !quantum.bit
+    // CHECK: [[_19:%.+]]:2 = quantum.custom "CNOT"() [[MINUS_0]], [[ONE_1]] : !quantum.bit, !quantum.bit
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[_19]]#0, [[_19]]#1 : !quantum.bit, !quantum.bit
+
+
+    %20:2 = quantum.custom "CNOT"() %OTHERS_0, %OTHERS_1 : !quantum.bit, !quantum.bit
+    %user_20:2 = quantum.custom "CRZ"(%cst) %20#0, %20#1 : !quantum.bit, !quantum.bit
+    // CHECK: [[_20:%.+]]:2 = quantum.custom "CNOT"() [[OTHERS_0]], [[OTHERS_1]] : !quantum.bit, !quantum.bit
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[_20]]#0, [[_20]]#1 : !quantum.bit, !quantum.bit
+
+
+    %21:2 = quantum.custom "CNOT"() %OTHERS_0, %PLUS_1 : !quantum.bit, !quantum.bit
+    %user_21:2 = quantum.custom "CRZ"(%cst) %21#0, %21#1 : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.custom "CNOT"
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[OTHERS_0]], [[PLUS_1]] : !quantum.bit, !quantum.bit
+
+
+    %22:2 = quantum.custom "CNOT"() %OTHERS_0, %MINUS_1 : !quantum.bit, !quantum.bit
+    %user_22:2 = quantum.custom "CRZ"(%cst) %22#0, %22#1 : !quantum.bit, !quantum.bit
+    // CHECK-NOT: quantum.custom "CNOT"
+    // CHECK: [[mid_22:%.+]] = quantum.custom "PauliZ"() [[OTHERS_0]] : !quantum.bit
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[mid_22]], [[MINUS_1]] : !quantum.bit, !quantum.bit
+
+
+    %23:2 = quantum.custom "CNOT"() %OTHERS_0, %ZERO_1 : !quantum.bit, !quantum.bit
+    %user_23:2 = quantum.custom "CRZ"(%cst) %23#0, %23#1 : !quantum.bit, !quantum.bit
+    // CHECK: [[_23:%.+]]:2 = quantum.custom "CNOT"() [[OTHERS_0]], [[ZERO_1]] : !quantum.bit, !quantum.bit
+    // CHECK: {{%.+}} = quantum.custom "CRZ"({{%.+}}) [[_23]]#0, [[_23]]#1 : !quantum.bit, !quantum.bit
+
+
+    %24:2 = quantum.custom "CNOT"() %OTHERS_0, %ONE_1 : !quantum.bit, !quantum.bit
+    %user_24:2 = quantum.custom "CRZ"(%cst) %24#0, %24#1 : !quantum.bit, !quantum.bit
+    // CHECK: [[_24:%.+]]:2 = quantum.custom "CNOT"() [[OTHERS_0]], [[ONE_1]] : !quantum.bit, !quantum.bit
```

Comment:
This is a long chain of checks with always the same two gates involved, would you mind explaining the thinking behind it?

Reply:
So this is my test design:

Looking at the CNOT decomposition table (https://arxiv.org/pdf/2012.07711, page 5 table 1), I would like to test all 25 entries. To do this, I need to produce a bunch of qubit values whose state are known from the state machine. This is what I did with the big pool of qubit values at the top.

Then, I reuse these pool of known-state values for each entry. Each entry has two qubits, with known states in the FSM, going into the CNOT, and then going into a user gate. I just chose CRZ to be the user gate because (I don't know why). A user is necessary because I want to test whether the input qubits to the CNOT are properly rewired into the user, with possibly other gates inserted in between.

The reason why I reuse the same pool of known-state values is because otherwise if we split each table entry into its own test, every test would need to have this production line at the top for the known-states, which is very repetitive. For example, 4 lines are needed to produce a single |-> state (alloc, extract, gate(X), gate(H)).